ui_test/src/lib.rs

```diff
@@ -293,50 +316,111 @@ fn check_test_result(
         comments,
     );
     // Check error annotations in the source against output
-    check_annotations(&stderr, errors, config, revision, comments);
+    check_annotations(diagnostics, path, errors, config, revision, comments);
 }
 
 fn check_annotations(
-    unnormalized_stderr: &str,
+    mut diagnostics: Diagnostics,
+    path: &Path,
     errors: &mut Errors,
     config: &Config,
     revision: &str,
     comments: &Comments,
 ) {
-    let mut found_annotation = false;
+    let mut highest_annotation_level = Level::Error;
     if let Some((ref error_pattern, definition_line)) = comments.error_pattern {
-        if !unnormalized_stderr.contains(error_pattern) {
+        let mut found = false;
+
+        // first check the diagnostics messages outside of our file. We check this first, so that
+        // you can mix in-file annotations with // error-pattern annotations, even if there is overlap
+        // in the messages.
+        if let Some(i) = diagnostics
+            .messages_from_unknown_file_or_line
+            .iter()
+            .position(|msg| msg.message.contains(error_pattern))
+        {
+            diagnostics.messages_from_unknown_file_or_line.remove(i);
+            found = true;
+        }
+
+        // if nothing was found, check the ones inside our file. We permit this because some tests may have
+        // flaky line numbers for their messages.
+        if !found {
+            for line in &mut diagnostics.messages {
+                if let Some(i) = line.iter().position(|msg| msg.message.contains(error_pattern)) {
+                    line.remove(i);
+                    found = true;
+                    break;
+                }
+            }
+        }
+
+        if !found {
             errors.push(Error::PatternNotFound {
-                stderr: unnormalized_stderr.to_string(),
+                stderr: diagnostics.rendered.to_string(),
                 pattern: error_pattern.to_string(),
                 definition_line,
             });
         }
-        found_annotation = true;
     }
-    for &ErrorMatch { ref matched, revision: ref rev, definition_line } in &comments.error_matches {
-        // FIXME: check that the error happens on the marked line
-
+    for &ErrorMatch { ref matched, revision: ref rev, definition_line, line, level } in
+        &comments.error_matches
+    {
         if let Some(rev) = rev {
             if rev != revision {
                 continue;
             }
         }
+        if let Some(level) = level {
+            // If we found a diagnostic with a level annotation, make sure that all
+            // diagnostics of that level have annotations, even if we don't end up finding a matching diagnostic
+            // for this pattern.
+            highest_annotation_level = std::cmp::max(highest_annotation_level, level);
```

Comment:
Please explain that the order here is such that "error" is low (which is the opposite of the order I would have expected).

Or maybe it'd be better to invert the order, and use `min` and `lowest_annotation-level`?